src/algebra/category/Group/epi_mono.lean

```diff
@@ -0,0 +1,334 @@
+by rw [tau, equiv.symm_swap, equiv.swap_apply_right]
+
+/--
+Let `g : B ⟶ S(X')` be defined as as such that, for any `β : B`, `g(β)` is the function sending
+point at infinity to point at infinity and sending coset `y` to `β *l y`.
+-/
+def G : B →* SX' :=
+{ to_fun := λ β,
+  { to_fun := λ x, β • x,
+    inv_fun := λ x, β⁻¹ • x,
+    left_inv := λ x, by dsimp only; rw [←mul_smul, mul_left_inv, one_smul],
+    right_inv := λ x, by dsimp only; rw [←mul_smul, mul_right_inv, one_smul] },
+  map_one' := by ext; simp [one_smul],
+  map_mul' := λ b1 b2, by ext; simp [mul_smul] }
+
+local notation `g` := G f
+
+/--
+Define `h : B ⟶ S(X')` to be `τ g τ⁻¹`
+-/
+noncomputable def H : B →* SX':=
+{ to_fun := λ β, ((τ).symm.trans (g β)).trans τ,
+  map_one' := by ext; simp,
+  map_mul' := λ b1 b2, by ext; simp }
+
+local notation `h` := H f
+
+/--
+The strategy is the following: assuming `epi f`
+* prove that `f.range = {x | h x = g x}`;
+* thus `f ≫ h = f ≫ g` so that `h = g`;
+* but if `f` is not surjective, then some `x ∉ f.range`, then `h x ≠ g x` at the coset `f.range`.
+-/
+
+lemma g_apply_from_coset (x : B) (y : X) :
+  (g x) (from_coset y) =
+  from_coset ⟨x *l y, begin
+    rw [←subtype.val_eq_coe, ←y.2.some_spec, set.mem_range, left_coset_assoc],
+    use x * y.2.some,
+  end⟩ := rfl
+
+lemma g_apply_infinity (x : B) :
+  (g x) ∞ = ∞ := rfl
+
+lemma h_apply_infinity (x : B) (hx : x ∈ f.range) :
+  (h x) ∞ = ∞ :=
+begin
+  simp only [H, monoid_hom.coe_mk, equiv.to_fun_as_coe, equiv.coe_trans, function.comp_app],
+  rw [τ_symm_apply_infinity, g_apply_from_coset],
+  simpa only [←subtype.val_eq_coe] using τ_apply_from_coset' f x hx,
+end
+
+lemma h_apply_from_coset (x : B) :
+  (h x) (from_coset ⟨f.range.carrier, ⟨1, one_left_coset _⟩⟩) =
+  from_coset ⟨f.range.carrier, ⟨1, one_left_coset _⟩⟩ :=
+by simp [H, τ_symm_apply_from_coset, g_apply_infinity, τ_apply_infinity]
+
+lemma h_apply_from_coset' (x : B) (b : B) (hb : b ∈ f.range):
+  (h x) (from_coset ⟨b *l f.range.carrier, ⟨b, rfl⟩⟩) =
+  from_coset ⟨b *l f.range.carrier, ⟨b, rfl⟩⟩ :=
+(from_coset_eq_of_mem_range _ hb).symm ▸ h_apply_from_coset f x
+
+lemma h_apply_from_coset_nin_range (x : B) (hx : x ∈ f.range) (b : B) (hb : b ∉ f.range) :
+  (h x) (from_coset ⟨b *l f.range.carrier, ⟨b, rfl⟩⟩) =
+  from_coset ⟨(x * b) *l f.range.carrier, ⟨x * b, rfl⟩⟩ :=
+begin
+  simp only [H, tau, monoid_hom.coe_mk, equiv.to_fun_as_coe, equiv.coe_trans, function.comp_app],
+  rw [equiv.symm_swap],
+  rw @equiv.swap_apply_of_ne_of_ne X' _
+    (from_coset ⟨f.range.carrier, ⟨1, one_left_coset _⟩⟩) ∞
+    (from_coset ⟨b *l f.range.carrier, ⟨b, rfl⟩⟩) (from_coset_ne_of_nin_range _ hb) (by simp),
+  simp only [g_apply_from_coset, ←subtype.val_eq_coe, left_coset_assoc],
+  exact equiv.swap_apply_of_ne_of_ne begin
+    refine from_coset_ne_of_nin_range _ (λ r, hb _),
+    convert subgroup.mul_mem _ (subgroup.inv_mem _ hx) r,
+    rw [←mul_assoc, mul_left_inv, one_mul],
+  end (by simp),
```

Comment:
We try to avoid `begin..end` blocks inside of tactic proofs. It's usually easier to read if you create new goals instead:
```suggestion
  refine equiv.swap_apply_of_ne_of_ne (from_coset_ne_of_nin_range _ (λ r, hb _)) (by simp),
  convert subgroup.mul_mem _ (subgroup.inv_mem _ hx) r,
  rw [←mul_assoc, mul_left_inv, one_mul],
```